# Elementary programming

Elementary programming is the small set of operations every program needs before it can
do anything interesting: read a value in, store it under a name, compute with operators,
and display the result.

Everything in this lesson fits that four-step shape.

    input                store               compute                  output
    -----------------    ----------------    --------------------    -----------------
    input()          ->  variable name   ->  + - * / // % **     ->  print()
    a literal            named constant      int(), round()          f"{value:.2f}"

Code cells are meant to be edited and re-run: change a value, predict the result, then
run the cell and compare.

# Simple program

A program is a sequence of statements the interpreter executes in order, top to bottom.
A statement is one instruction, such as an assignment or a call to `print`. An
expression is anything that produces a value, such as `100 * 0.0825`. Statements are
where values get stored or displayed; expressions are what produce them.

Two habits before writing any code:

- **Comments.** Everything after `#` on a line is ignored by the interpreter. A comment
  records intent, not mechanics: `# 8.25% state rate` is useful, `# multiply` is not.
- **Pseudocode.** Pseudocode is the algorithm in plain language with a little code mixed
  in. Write it first, then translate it line by line:

      read the bill amount
      set the tax rate to 8.25%
      tax = bill amount * tax rate
      total = bill amount + tax
      display tax and total

The cell below is that pseudocode turned into Python.

In [1]:
# Enter bill amount
bill_amount = 100  # bill is now 100

# Define tax rate
tax_rate = 0.0825  # 8.25%

# Compute tax
tax = bill_amount * tax_rate

# Compute total bill
total_bill = bill_amount + tax

# Display results
print("Tax:", tax)
print("Total bill:", total_bill)

Tax: 8.25
Total bill: 108.25


## Tracing execution

Tracing means stepping through the statements by hand and writing down what each
variable holds after each one. It is how you find a wrong answer that does not raise an
error.

| Statement | `bill_amount` | `tax_rate` | `tax` | `total_bill` |
|---|---|---|---|---|
| `bill_amount = 100` | 100 | does not exist | does not exist | does not exist |
| `tax_rate = 0.0825` | 100 | 0.0825 | does not exist | does not exist |
| `tax = bill_amount * tax_rate` | 100 | 0.0825 | 8.25 | does not exist |
| `total_bill = bill_amount + tax` | 100 | 0.0825 | 8.25 | 108.25 |

A variable does not exist until a statement assigns to it, each new assignment replaces
the previous value, and nothing appears on screen until `print` runs.

In [20]:
# Same three steps, printed after each one so the state is visible.
credit_hours = 12
print("after statement 1: credit_hours =", credit_hours)

cost_per_hour = 275.50
print("after statement 2: cost_per_hour =", cost_per_hour)

tuition = credit_hours * cost_per_hour
print("after statement 3: tuition       =", tuition)

after statement 1: credit_hours = 12
after statement 2: cost_per_hour = 275.5
after statement 3: tuition       = 3306.0


> **Assignment**:  
> _______ is the code in natural language mixed with some program code.  
>  
> - [ ] A flowchart diagram  
> - [ ] A Python statement  
> - [x] Pseudocode  
> - [ ] Python program  

# Reading input from console

`input` displays a prompt, waits for the user to type a line, and returns everything
typed as a string. It never returns a number, even when the user types digits.

    text   = input("Enter a string: ")          # always a str
    count  = int(input("Enter an integer: "))   # converted to int
    amount = float(input("Enter a number: "))   # converted to float

Converting is a separate step, and it fails loudly: `int("twelve")` raises `ValueError`.
Wrap the call in `int()` or `float()` as soon as the value is read, so the rest of the
program works with numbers rather than text.

In [2]:
tax_rate = input("Enter a tax rate in your state: ")
tax_rate

Enter a tax rate in your state:  8.5


'8.5'

The cell above returned `'8.5'` with quotes around it: a string, not a number. The
difference matters immediately, because the arithmetic operators do something else
entirely to strings. `'8.5' * 3` repeats the text three times instead of scaling the
value, and `'8.5' + 1` raises `TypeError`.

In [21]:
typed = "8.5"          # stands in for what input() returned above
print(typed, type(typed))

rate = float(typed)    # now it is a number
print(rate, type(rate))

print("string * 3:", typed * 3)   # repeats the text
print("number * 3:", rate * 3)    # scales the value

8.5 <class 'str'>
8.5 <class 'float'>
string * 3: 8.58.58.5
number * 3: 25.5


> **Assignment**:  
> What function do you use to read a string?  
>  
> - [ ] int(enter("Enter a string"))  
> - [ ] float(enter("Enter a string"))  
> - [ ] int(input("Enter a string"))  
> - [ ] enter("Enter a string")  
> - [x] input("Enter a string")  

> **Assignment**:  
> If you enter 1 2 3 in three separate lines, when you run this program, what will be displayed?  
>  
> ```python
> print("Enter three numbers: ")
> number1 = int(input())
> number2 = int(input())
> number3 = int(input())
>
> # Compute average
> average = (number1 + number2 + number3) / 3
> 
> # Display result
> print(average)
> ```
> 
> - [x] 2.0  
> - [ ] 4.0  
> - [ ] 1.0  
> - [ ] 3.0  

In [22]:
# Reading three exam scores would be:
#   score1 = int(input("Score 1: "))
#   score2 = int(input("Score 2: "))
#   score3 = int(input("Score 3: "))
# Hard-coded here so the cell re-runs without typing.
score1, score2, score3 = 88, 94, 79

average = (score1 + score2 + score3) / 3
print("Average:", average)

Average: 87.0


> **Assignment**:  
> You can place the line continuation symbol __ at the end of a line to tell the interpreter that the statement is continued on the next line.  
>  
> - [ ] `/`  
> - [ ] `*`  
> - [ ] `&`  
> - [ ] `#`  
> - [x] `\`  

In [4]:
tax_rate = input("Enter a tax rate in your \
state: ")

Enter a tax rate in your state:  8.5


## Splitting a long statement across lines

A statement normally ends at the end of the line. Two ways to continue it:

- A backslash as the last character on the line, as in the cell above. Nothing may
  follow it, not even a space, which makes it easy to break by accident.
- Any unclosed `(`, `[`, or `{`. Inside brackets, Python keeps reading until they close,
  and indentation inside is free. This is the form to prefer.

In [23]:
# Implicit continuation inside parentheses: no backslash needed.
total_points = (
    88 + 94 + 79
    + 91 + 85
)
print("total points:", total_points)

# A long print call split the same way.
print("Average of five scores:",
      total_points / 5)

total points: 437
Average of five scores: 87.4


# Identifiers

An identifier is the name you give a variable, function, class, or module. The rules:

- It is made of letters, digits, and underscores (`_`) only. No spaces, no hyphens, no
  operators. `*` is multiplication and is never part of a name.
- It cannot start with a digit. `score1` is legal, `1score` is not.
- It cannot be a reserved word. Reserved words, also called keywords, already have a
  meaning in the language.
- It has no length limit, and it is case sensitive: `total`, `Total`, and `TOTAL` are
  three different names.

Conventions layered on top of the rules: `lower_snake_case` for variables and functions,
`UPPER_SNAKE_CASE` for constants, and a leading underscore for names meant to stay
internal to a module. A name that describes the value, such as `credit_hours`, is worth
more than a short one, such as `ch`.

In [24]:
import keyword

print(len(keyword.kwlist), "reserved words:")
print(keyword.kwlist)

print("is 'class' a keyword?", keyword.iskeyword("class"))
print("is 'grade' a keyword?", keyword.iskeyword("grade"))

35 reserved words:
['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']
is 'class' a keyword? True
is 'grade' a keyword? False


In [25]:
# Case sensitivity: three separate variables.
total = 10
Total = 20
TOTAL = 30
print(total, Total, TOTAL)

# An illegal name is a syntax error, caught before anything runs:
#   1score = 5      -> SyntaxError: invalid decimal literal
#   class = "BCIS"  -> SyntaxError: invalid syntax

10 20 30


> **Assignment**:  
> An identifier cannot be a keyword?  
>  
> - [x] true  
> - [ ] false  

> **Assignment**:  
> An identifier can contain digits, but cannot start with a digit?  
>  
> - [x] true  
> - [ ] false  

> **Assignment**:  
> Which of the following is a valid identifier?  
>  
> - [x] mile  
> - [ ] 8+9  
> - [ ] 9X  
> - [ ] $343  

# Variables, expressions, and assignment

A variable is a name bound to a value. The assignment statement

    name = expression

evaluates the expression on the right first, then binds the name on the left to the
result. That order is what explains `x = x + 1`: Python computes the current value of
`x` plus one, and only then rebinds `x`. Read `=` as "becomes", not as "equals".

Four consequences worth stating:

- A name must be assigned before it is used. Using an unassigned name raises
  `NameError`, which is what the next few cells demonstrate.
- Assigning again replaces the binding, and the previous value is gone.
- The name carries no type of its own. `points = 3` followed by `points = 3.5` is legal;
  the type comes from the value.
- `i = j = k = 1` is a chained assignment: one expression, three names bound to it.

In [26]:
enrolled = 24
enrolled = enrolled + 3      # right side first (27), then rebind
print("enrolled:", enrolled)

first = second = third = 0   # chained assignment
print(first, second, third)

points = 3                   # the name does not fix the type
print(points, type(points))
points = 3.5
print(points, type(points))

enrolled: 27
0 0 0
3 <class 'int'>
3.5 <class 'float'>


> **Goal**: Learn to assign variables with different names and types.  
> **Assignment**: Assign to the variable `eightBelow` the value `-8`.

In [9]:
eightBelow = -8

> **Goal**: Learn to assign variables with different names and types.  
> **Assignment**: Assign to the variable `goldenRatio` an approximation of the value of the most important number in mathematics, `1.618034`.

In [10]:
goldenRatio = 1.618034

> **Goal**: Learn to assign variables with different names and types  
> **Assignment**: Assign to the variable `averageTemperature` the value `-454.8` and to the variable `speed` the value `41.9`.

In [11]:
averageTemperature = -454.8
speed = 41.9

> **Assignment**: Analyze the following code:
>
> ```python
> interestRate = 0.05
> interest = interestrate * 45
> ```
>
> - [ ] interest is not defined.  
> - [ ] The code is missing a semicolon at the end of the statement.  
> - [x] interestrate is not defined.  
> - [ ] interestRate is not defined.  

In [12]:
interestRate = 0.05
interest = interestrate * 45

NameError: name 'interestrate' is not defined

> **Assignment**: What will be displayed by the following code?
> 
> ```python
> x = 1
> x = 2 * x + 1
> print(x)
> ```
> 
> - [ ] 2  
> - [ ] 1  
> - [ ] 0  
> - [x] 3  
> - [ ] 4  

In [14]:
x = 1
x = 2 * x + 1
print(x)

3


> **Assignment**: What will be displayed by the following code?
> 
> ```python
> x = 1
> x = x + 2.5
> print(x)
> ```
> 
> - [ ] The statements are illegal  
> - [ ] 3  
> - [ ] 1  
> - [x] 3.5  
> - [ ] 2  

In [15]:
x = 1
x = x + 2.5
print(x)

3.5


# Simultaneous assignment

`var1, var2, ..., varn = exp1, exp2, ..., expn`

Simultaneous assignment evaluates every expression on the right side before binding any
name on the left. The right side is packed into a tuple, then unpacked into the names
from left to right.

That is why swapping needs no temporary variable: `y, x` is built from the old values,
so `x, y = y, x` cannot overwrite a value it still needs. The counts on the two sides
must match, otherwise Python raises `ValueError`.

In [27]:
midterm, final = 82, 91
print("before:", midterm, final)

midterm, final = final, midterm    # both right-side values read first
print("after: ", midterm, final)

try:
    a, b, c = 1, 2                 # three names, two values
except ValueError as error:
    print("ValueError:", error)

before: 82 91
after:  91 82
ValueError: not enough values to unpack (expected 3, got 2)


In [17]:
x = 1
y = 0
print(x, y)

1 0


In [18]:
x, y = y, x
print(x, y)

0 1


> **Assignment**: Write a statement that swaps the values of `val` and `temp`.  

```python
val, temp = temp, val
```

> **Assignment**: Given the variable `catAge`, write a statement that assigns to `humanAge` four times the value of `catAge`.  

```python
humanAge = 4 * catAge
```

> **Assignment**: What will be displayed by the following code?
> 
> ```python
> x, y = 1, 2
> x, y = y, x
> print(x, y)
> ```
> 
> - [ ] 1 1  
> - [x] 2 1  
> - [ ] 2 2  
> - [ ] 1 2  

In [19]:
x, y = 1, 2
x, y = y, x
print(x, y)

2 1


# Named Constant

`named constant` - an identifier for a permanent value.

A named constant is a variable you promise not to reassign. Python has no `const`
keyword and will not stop you from changing it; the all-uppercase name is the signal to
every reader that the value is fixed.

Two reasons to name a number instead of typing it inline: the name says what the number
means, and the value appears in exactly one place, so changing a rate is a one-line edit
rather than a search through the program.

In [3]:
TEXAS_TAX_RATE = 8.25 / 100
TEXAS_TAX_RATE

0.0825

In [28]:
TEXAS_TAX_RATE = 0.0825
MONTHS_PER_YEAR = 12

# The constants carry the meaning, so the arithmetic reads as arithmetic.
textbook_price = 189.00
print("tax on the textbook:", textbook_price * TEXAS_TAX_RATE)

annual_fee = 1200.00
print("monthly share of the fee:", annual_fee / MONTHS_PER_YEAR)

tax on the textbook: 15.592500000000001
monthly share of the fee: 100.0


# Numeric data types

Two numeric types carry every value in this lesson.

| Type | Written as | What it holds |
|---|---|---|
| `int` | `3`, `-8`, `1_000_000` | whole numbers, no fractional part |
| `float` | `3.0`, `-0.5`, `2.5e3` | numbers with a fractional part |

`int` has no fixed width. It grows to as many digits as memory allows, so integer
arithmetic never wraps around and never loses precision.

`float` stores an approximation of a decimal value in a fixed 64 bits, in binary. Most
decimal fractions have no exact binary form, so tiny representation errors show up in
results: `0.1 + 0.2` prints as `0.30000000000000004`. This is a property of the storage
format rather than a Python defect, and it is why money is displayed with formatting and
compared with a tolerance instead of tested for exact equality.

Mixing the two promotes the result to `float`, so `7 + 0.5` is `7.5` and `4 / 2` is
`2.0`. Underscores may be used inside a numeric literal as digit separators; the
interpreter ignores them.

In [29]:
print(type(3), type(3.0))

print("2 ** 200 =", 2 ** 200)                 # int grows to fit

print("0.1 + 0.2       =", 0.1 + 0.2)         # float approximation is visible
print("0.1 + 0.2 == 0.3 ->", 0.1 + 0.2 == 0.3)

print("7 + 0.5 =", 7 + 0.5, type(7 + 0.5))    # int + float promotes to float

<class 'int'> <class 'float'>
2 ** 200 = 1606938044258990275541962092341162602522202993782792835301376
0.1 + 0.2       = 0.30000000000000004
0.1 + 0.2 == 0.3 -> False
7 + 0.5 = 7.5 <class 'float'>


# Numeric Operators

| Name | Meaning           | Example        | Result |
|------|-------------------|----------------|--------|
| `+`  | Addition          | `34 + 1`       | `35`   |
| `-`  | Subtraction       | `34.0 - 0.1`   | `33.9` |
| `*`  | Multiplication    | `300 * 30`     | `9000` |
| `/`  | Float division    | `1 / 2`        | `0.5`  |
| `//` | Integer division  | `1 // 2`       | `0`    |
| `**` | Exponentiation    | `4 ** 0.5`     | `2.0`  |
| `%`  | Remainder         | `20 % 3`       | `2`    |

## Two kinds of quotient, and exponentiation

`/` always produces a `float`, even when the division comes out even: `4 / 2` is `2.0`.

`//` produces the floor of the quotient, the largest whole number less than or equal to
the true result. With positive operands that looks like truncation, but with a negative
operand it moves away from zero: `-7 // 2` is `-4`, not `-3`. The type follows the
operands, so two ints give an int and any float gives a float.

`**` raises to a power and accepts a fractional exponent, which makes `4 ** 0.5` the
square root `2.0`.

In [30]:
print("4 / 2    =", 4 / 2)        # float division, always a float
print("7 / 2    =", 7 / 2)
print("7 // 2   =", 7 // 2)       # floor division, int result
print("-7 // 2  =", -7 // 2)      # floors downward, not toward zero
print("7.0 // 2 =", 7.0 // 2)     # a float operand gives a float result
print("4 ** 0.5 =", 4 ** 0.5)     # fractional exponent is a root
print("2 ** 10  =", 2 ** 10)

4 / 2    = 2.0
7 / 2    = 3.5
7 // 2   = 3
-7 // 2  = -4
7.0 // 2 = 3.0
4 ** 0.5 = 2.0
2 ** 10  = 1024


In [8]:
print(100000000)
print(100_000_000)

100000000
100000000


> **Assignment:**  
Given the variable `taxablePrice`, write a statement that assigns to `fullPrice` the value of `taxablePrice` plus the sales tax of 25%.

```python
fullPrice = taxablePrice * 1.25
```

> **Assignment:**  
By policy, a store only bills its loyal clients for the whole part of their purchases (e.g., $5.70 is billed as $5.00—any fractional part is truncated). Assuming the price is stored in the variable `fullPrice`, write code that assigns to `billedPrice` the whole part of `fullPrice`.

```python
billedPrice = int(fullPrice)
```

> **Assignment:**  
Starting with two rabbits, a colony doubles its members at each generation. The first generation will count two rabbits, the second four, the third eight, and so on. Assuming the generation is assigned to the variable `generation`, write a statement that assigns to `totalRabbits` the number of rabbits in the colony.

```python
totalRabbits = 2 ** generation
```

...

# Modulo operator

`%` returns what is left after floor division. For any `a` and any non-zero `b`:

    a == (a // b) * b + (a % b)

Three standard uses:

1. **Testing divisibility.** `n % 2` is `0` for an even number and `1` for an odd one.
   `n % 15 == 0` tests divisibility by 15.
2. **Splitting a total into units.** Total minutes into hours and minutes:
   `hours = total // 60` and `minutes = total % 60`.
3. **Wrapping around a cycle.** With the seven weekdays numbered 0 to 6, the day 10 days
   after day 6 is `(6 + 10) % 7`, which is day 2.

The result takes the sign of the right operand, so `-7 % 3` is `2`, not `-1`.

In [31]:
class_size = 23
print("can the class split into pairs?", class_size % 2 == 0)

total_minutes = 275
print(total_minutes, "minutes =", total_minutes // 60, "h", total_minutes % 60, "min")

SATURDAY = 6
print("10 days after day 6 is day", (SATURDAY + 10) % 7)

print("-7 % 3 =", -7 % 3)                          # sign follows the divisor
print("identity check:", (-7 // 3) * 3 + (-7 % 3))  # back to -7

can the class split into pairs? False
275 minutes = 4 h 35 min
10 days after day 6 is day 2
-7 % 3 = 2
identity check: -7


> **Assignment:**  
A teacher collects candies for each time a student answers a question correctly during the week. At the end of the week, they distribute the candy equally across the classroom. If any candy is left over, it stays in the candy jar for the week after. Assuming `classSize` stores the number of students in the class, and `candies` the number of candies collected, write a statement that assigns to `leftover` the candies that remain in the jar at the end of the week.

```python
leftover = candies % classSize
```

> **Assignment**: Which of the following expressions results in a value of `1`?
> 
> - [ ] `15 % 4`
> - [x] `37 % 6`
> - [ ] `25 % 5`
> - [ ] `2 % 1`

> **Assignment**: 25 % 1 is _____
> 
> - [ ] 1
> - [x] 0
> - [ ] 2
> - [ ] 3
> - [ ] 4

> **Assignment**: 24 % 5 is _____
> 
> - [ ] 2
> - [ ] 0
> - [ ] 3
> - [ ] 1
> - [x] 4

# Scientific notation

Scientific notation writes a float as a mantissa times a power of ten. Python writes the
power with `e` or `E`, in either case.

| Literal | Means | Value |
|---|---|---|
| `1.23456e2` | 1.23456 x 10^2 | 123.456 |
| `1.23456e+2` | the same; the `+` is optional | 123.456 |
| `1.23456e-2` | 1.23456 x 10^-2 | 0.0123456 |

A literal written this way is a `float` even when the value is whole, so `2e3` is
`2000.0`. Python also switches to this form on output once a value is large or small
enough that the decimal form would be unreadable.

In [32]:
print(1.23456e2, 1.23456e+2, 1.23456e-2)
print("2e3 is", 2e3, type(2e3))

print(0.0001)     # still printed as a decimal
print(0.00001)    # switches to e-notation
print(2 ** 100 * 1.0)

123.456 123.456 0.0123456
2e3 is 2000.0 <class 'float'>
0.0001
1e-05
1.2676506002282294e+30


> **Assignment**: Which of the following is not equivalent to 0.025?
> 
> - [ ] 0.25E-1
> - [ ] 0.00025E2
> - [x] 0.0025E+2
> - [ ] 2.5e-2
> - [ ] 0.0025E1

- **E-notation** means “× 10 to the power of”.  
- Only `0.0025E+2` evaluates to **0.25**, not **0.025**, so it’s the one **not** equivalent.  

**Explanation of each expression**:

| Expression     | Interpretation               | Value   | Equivalent to 0.025? |
|---------------:|------------------------------|--------:|:--------------------:|
| `0.25E-1`      | 0.25 × 10^(–1)               | 0.025   | Yes                  |
| `0.00025E2`    | 0.00025 × 10^2               | 0.025   | Yes                  |
| `0.0025E+2`    | 0.0025 × 10^2                | 0.25    | **No**               |
| `2.5e-2`       | 2.5 × 10^(–2)                | 0.025   | Yes                  |
| `0.0025E1`     | 0.0025 × 10^1                | 0.025   | Yes                  |

- **E-notation** means “× 10 to the power of”.  
- Only `0.0025E+2` evaluates to **0.25**, not **0.025**, so it’s the one **not** equivalent.  

# Overflow and underflow

Overflow happens when a result is too large for the type that must hold it.

`int` has no fixed size, so integers do not overflow: `245 ** 1000` returns an exact
whole number with thousands of digits.

`float` is fixed at 64 bits, with a largest representable value near 1.8 x 10^308. A
float computation past that limit raises `OverflowError`, while arithmetic on values
that are already floats gives `inf`, the floating-point infinity, instead of raising.

Underflow is the opposite end: a float too close to zero to represent. Python
approximates it to `0.0` without complaint, which is usually harmless. The one thing to
watch is dividing by a value that has underflowed, since that raises
`ZeroDivisionError`.

In [33]:
import sys

big = 245 ** 1000                       # int: exact, no size limit
print("digits in 245 ** 1000:", len(str(big)))

print("largest float:", sys.float_info.max)

try:
    245.0 ** 1000                       # float: past the limit
except OverflowError as error:
    print("OverflowError:", error)

print("1e308 * 10   =", 1e308 * 10)     # existing floats go to inf
print("1e-320 / 1e10 =", 1e-320 / 1e10)  # underflow, approximated to 0.0

digits in 245 ** 1000: 2390
largest float: 1.7976931348623157e+308
OverflowError: (34, 'Result too large')
1e308 * 10   = inf
1e-320 / 1e10 = 0.0


> **Assignment**: If a number is too large to be stored in memory, it _____________.
> 
> - [ ] causes underflow
> - [ ] cannot happen in Python
> - [ ] causes no error
> - [x] causes overflow

# Arithmetic expressions and precedence

Translating an algebraic expression into Python means writing out two things that
mathematical notation leaves implicit: every multiplication needs an explicit `*`, and
every fraction bar becomes a `/` with parentheses around whatever was above and below it.

    3 + 4x
    ------ + 10(y - 5)     becomes     (3 + 4 * x) / 5 + 10 * (y - 5)
       5

Python then evaluates the expression by the ordinary arithmetic rules.

| Order | Operators | Grouping |
|---|---|---|
| 1 | `()` | innermost pair first |
| 2 | `**` | right to left |
| 3 | unary `-` | |
| 4 | `*` `/` `//` `%` | left to right |
| 5 | `+` `-` | left to right |

Two rows of that table cause most wrong answers. `**` binds tighter than everything
else, so `2 + 2 ** 3 / 2` is `2 + 8 / 2`, which is `6.0`. And `**` groups right to left,
so `2 ** 3 ** 2` is `2 ** 9`, not `8 ** 2`. When the intended order is not obvious at a
glance, add the parentheses even where they are redundant.

In [34]:
x, y = 2, 7
print("(3 + 4 * x) / 5 + 10 * (y - 5) =", (3 + 4 * x) / 5 + 10 * (y - 5))

print("2 + 2 ** 3 / 2 =", 2 + 2 ** 3 / 2)   # ** first, then /, then +
print("2 ** 3 ** 2    =", 2 ** 3 ** 2)      # right to left: 2 ** 9
print("(2 ** 3) ** 2  =", (2 ** 3) ** 2)
print("-2 ** 2        =", -2 ** 2)          # ** binds tighter than unary minus

(3 + 4 * x) / 5 + 10 * (y - 5) = 22.2
2 + 2 ** 3 / 2 = 6.0
2 ** 3 ** 2    = 512
(2 ** 3) ** 2  = 64
-2 ** 2        = -4


> **Assignment:**  
> Given the variables `temperature1`, `temperature2`, and `temperature3`, write a single statement that assigns to `averageTemperature` the average of the three temperatures.
```python
averageTemperature = (temperature1 + temperature2 + temperature3) / 3
```

> **Assignment:**  
> A house is made of four rooms. Rooms 1 and 2 are rectangular, while rooms 3 and 4 are right triangles. The measurements for room 1 are stored in `width1` and `length1`, for room 2 in `width2` and `length2`, for room 3 in `width3` and `length3`, and for room 4 in `width4` and `length4`. Note: the area of a rectangle is `width * length`, while for a right triangle it is `(width * length) / 2`. Write a single statement assigning to `totalSurface` the total surface area of the house.

```python
totalSurface = width1 * length1 + width2 * length2 + (width3 * length3) / 2 + (width4 * length4) / 2
```

> **Assignment:**  
> In mathematics, the alternating harmonic series is  
> 1 − 1/2 + 1/3 − 1/4 + 1/5 − … + ((-1)^(N+1)) / N.  
> Write a statement that assigns to `tenthAltHarmonic` the value of the tenth alternating harmonic number (i.e., 1 − 1/2 + 1/3 − … + ((-1)^(10+1)) / 10).
```python
tenthAltHarmonic = 1 - 1/2 + 1/3 - 1/4 + 1/5 - 1/6 + 1/7 - 1/8 + 1/9 - 1/10
```

> **Assignment**: What is the result of evaluating `2 + 2 ** 3 / 2`?
> 
> - [ ] 4.0
> - [x] 6.0
> - [ ] 6
> - [ ] 4


# Augmented Assignment Operators

The operators `+`, `-`, `*`, `/`, `//`, `%`, and `**` can be combined with the assignment operator (`=`) to form augmented assignment operators.

| Operator | Name                         | Example   | Equivalent       |
|----------|------------------------------|-----------|------------------|
| `+=`     | Addition assignment          | `i += 8`  | `i = i + 8`      |
| `-=`     | Subtraction assignment       | `i -= 8`  | `i = i - 8`      |
| `*=`     | Multiplication assignment    | `i *= 8`  | `i = i * 8`      |
| `/=`     | Float division assignment    | `i /= 8`  | `i = i / 8`      |
| `//=`    | Integer division assignment  | `i //= 8` | `i = i // 8`     |
| `%=`     | Remainder assignment         | `i %= 8`  | `i = i % 8`      |
| `**=`    | Exponent assignment          | `i **= 8` | `i = i ** 8`     |

An augmented assignment reads the variable, applies the operator, and rebinds the same
name. `i += 8` is `i = i + 8`, so the variable must already hold a value: `count += 1` on
an unassigned `count` raises `NameError`.

The operator applies to the whole expression on the right, not only to its first term.
`i *= j + 5` means `i = i * (j + 5)`, not `i = i * j + 5`.

In [35]:
graded = 0
to_grade = 30

graded += 5        # graded = graded + 5
to_grade -= 5      # to_grade = to_grade - 5
print("graded:", graded, " still to grade:", to_grade)

i, j = 2, 3
i *= j + 5         # i = i * (3 + 5), not i * 3 + 5
print("i:", i)

graded: 5  still to grade: 25
i: 16


In [ ]:
count = 1
count += 1

> **Assignment**: The number of attendees to an event is stored in `attendees`. Use an augmented assignment operator to write a single statement that increases the number of attendees by 7.
> 
```python
attendees += 7
```

> **Assignment**: Assume the two variables `graded` and `toGrade` store the number of graded assignments and the number still to be graded. Write some code that will grade two more assignments. That is, add 2 to `graded` and subtract 2 from `toGrade`. You must use augmented assignment operators in the statements.
```python
graded += 2
toGrade -= 2
```

> **Assignment**: Assume the two variables `graded` and `toGrade` store the number of graded assignments and the number still to be graded. Write some code that will grade two more assignments. That is, add 2 to `graded` and subtract 2 from `toGrade`. You must use augmented assignment operators in the statements.
```python
graded += 2
toGrade -= 2
```

> **Assignment**: What is the value of `i` printed?
> 
> ```python
> j = i = 1
> i += j + j * 5
> print("What is i?", i)
> ```
> 
> - [ ] 1
> - [ ] 6
> - [ ] 5
> - [x] 7
> - [ ] 0

> **Assignment**: What is `x` after the following statements?
> 
> ```python
> x = 1
> x *= x + 1
> ```
> 
> - [x] x is 2
> - [ ] x is 3
> - [ ] x is 4
> - [ ] x is 1

> **Assignment**: What is `x` after the following statements?
> 
> ```python
> x = 2
> y = 1
> x //= y + 1
> ```
> 
> - [ ] x is 2.  
> - [ ] x is 4.  
> - [x] x is 1.  
> - [ ] x is 3.  

# Rounding

Converting a number means calling the target type. `int(value)` and `float(value)` go
between the two numeric types, and `str(value)` turns a number into text so it can be
joined to a string with `+`.

`int()` truncates toward zero. It drops the fractional part rather than rounding, so
`int(4.9)` is `4` and `int(-4.9)` is `-4`. Use it when the whole part is what you want,
such as counting complete weeks in a term.

`round()` rounds to the nearest value and returns an `int` when called with one
argument. On an exact tie it rounds to the even neighbour, so `round(0.5)` is `0` and
`round(1.5)` is `2`. Called as `round(value, n)` it keeps `n` digits after the decimal
point and returns a `float`.

Rounding a float can still look wrong, because the value handed to `round` is already a
binary approximation: `round(2.675, 2)` gives `2.67`, since the stored value sits just
below 2.675. For displaying money, format the value rather than rounding it. The f-string
format `f"{amount:.2f}"` always produces exactly two decimal places, and it does not
change the number itself.

In [36]:
print("int(4.9)  =", int(4.9))       # truncates toward zero
print("int(-4.9) =", int(-4.9))
print("float(4)  =", float(4))
print("str(4) + '%' =", str(4) + "%")

print("round(3.4) =", round(3.4))
print("round(3.9) =", round(3.9))
print("round(0.5) =", round(0.5), " round(1.5) =", round(1.5))   # ties go to even
print("round(2.675, 2) =", round(2.675, 2))                      # binary approximation

purchase = 189.00
tax = purchase * 0.0825
print("raw total:      ", purchase + tax)
print("formatted total:", f"${purchase + tax:.2f}")

int(4.9)  = 4
int(-4.9) = -4
float(4)  = 4.0
str(4) + '%' = 4%
round(3.4) = 3
round(3.9) = 4
round(0.5) = 0  round(1.5) = 2
round(2.675, 2) = 2.67
raw total:       204.5925
formatted total: $204.59


> **Assignment**: Which of the following functions returns `4`?
> 
> - [x] `round(3.9)`
> - [ ] `int(3.9)`
> - [ ] `round(3.4)`
> - [ ] `int(3.4)`


# Reading the current time

`time.time()` returns the number of seconds elapsed since midnight, 1 January 1970,
Coordinated Universal Time (UTC). That reference point is called the epoch. The value is
a float, with fractions of a second after the decimal point, so it also serves as a
stopwatch: take it before and after a computation and subtract.

Turning that count into a clock time is nothing but `//` and `%`:

    total seconds since the epoch
        current second   = total seconds % 60
        total minutes    = total seconds // 60
        current minute   = total minutes % 60
        total hours      = total minutes // 60
        current UTC hour = total hours % 24

In [37]:
import time

current_time = time.time()
print("seconds since the epoch:", current_time)

total_seconds = int(current_time)
current_second = total_seconds % 60
total_minutes = total_seconds // 60
current_minute = total_minutes % 60
total_hours = total_minutes // 60
current_hour = total_hours % 24

# :02d pads a single digit with a leading zero, so 16:4:9 reads as 16:04:09.
print(f"{current_hour:02d}:{current_minute:02d}:{current_second:02d} UTC")

seconds since the epoch: 1790440910.866261
16:41:50 UTC


# The software development process

The software development process is the ordered set of phases that turns a problem
statement into working, maintained software. The phases are the same whether the result
is a 20-line class exercise or a system with a thousand users; only the amount of work in
each phase changes.

    requirement specification
              |
              v
        system analysis        input -> process -> output
              |
              v
         system design
              |
              v
        implementation
              |
              v
           testing
              |
              v
         deployment  ->  maintenance  ->  back to requirements

| Phase | What it produces |
|---|---|
| Requirement specification | A written statement of what the software must do, agreed with the people who will use it. |
| System analysis | The data flow: what goes in, what comes out, what the system has to know. |
| System design | The components, what each is responsible for, and how they relate. |
| Implementation | Code for each component, written, debugged, and working together. |
| Testing | Evidence that the code meets the requirement specification, ideally gathered by someone who did not write it. |
| Deployment | The software made available to its users. |
| Maintenance | Fixes and changes as the requirements and the environment move. |

The core of analysis and design is IPO: input, process, output. Name the inputs, name the
outputs, then work out the process that connects them. On a small program this takes two
minutes on paper, and it is what keeps the program from being rewritten three times.

## Worked example: monthly loan payment

IPO applied to a student loan.

- **Input:** the annual interest rate as a percent, the term in years, the loan amount.
- **Process:** convert the annual percent into a monthly rate, apply the payment formula,
  multiply by the number of payments to get the total.
- **Output:** the monthly payment and the total payment, each to two decimals.

The monthly payment on a fixed-rate loan is

    payment = amount * monthly_rate / (1 - 1 / (1 + monthly_rate) ** months)

The two conversions in the process step are where this program usually goes wrong. An
annual rate of 5.5 percent is `5.5 / 100 / 12` per month, and a 10-year term is `10 * 12`
payments, not 10.

In [38]:
MONTHS_PER_YEAR = 12
PERCENT = 100

# From input() in a real program:
#   annual_interest_rate = float(input("Annual interest rate in percent: "))
annual_interest_rate = 5.5
years = 10
loan_amount = 27000.00

monthly_rate = annual_interest_rate / PERCENT / MONTHS_PER_YEAR
months = years * MONTHS_PER_YEAR

monthly_payment = loan_amount * monthly_rate / (1 - 1 / (1 + monthly_rate) ** months)
total_payment = monthly_payment * months

print(f"Monthly payment: ${monthly_payment:.2f}")
print(f"Total payment:   ${total_payment:.2f}")
print(f"Interest paid:   ${total_payment - loan_amount:.2f}")

Monthly payment: $293.02
Total payment:   $35162.51
Interest paid:   $8162.51


## Worked example: distance between two points

The distance between the points (x1, y1) and (x2, y2) is

    ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5

The exponent `0.5` is the square root, so this needs nothing beyond the operators already
covered. The `math` module spells the same computation two clearer ways: `math.sqrt(v)`,
and `math.hypot(dx, dy)`, which takes the two differences directly and avoids overflow on
large coordinates. An `import` statement makes a module's contents available under its
name, and belongs at the top of the program.

In [39]:
import math

x1, y1 = 1.5, -3.4
x2, y2 = 4.0, 5.9

dx = x2 - x1
dy = y2 - y1

print("operators: ", (dx ** 2 + dy ** 2) ** 0.5)
print("math.sqrt: ", math.sqrt(dx ** 2 + dy ** 2))
print("math.hypot:", math.hypot(dx, dy))
print(f"rounded to two decimals: {math.hypot(dx, dy):.2f}")

operators:  9.630160954002795
math.sqrt:  9.630160954002795
math.hypot: 9.630160954002795
rounded to two decimals: 9.63
